src/Avalonia.Visuals/Media/DrawingContext.cs

```diff
@@ -2,23 +2,26 @@
 using System.Collections.Generic;
 using Avalonia.Media.Imaging;
 using Avalonia.Platform;
+using Avalonia.Threading;
 using Avalonia.Visuals.Media.Imaging;
 
 namespace Avalonia.Media
 {
     public sealed class DrawingContext : IDisposable
     {
+        private readonly bool _ownsImpl;
         private int _currentLevel;
 
 
-        static readonly Stack<Stack<PushedState>> StateStackPool = new Stack<Stack<PushedState>>();
-        static readonly Stack<Stack<TransformContainer>> TransformStackPool = new Stack<Stack<TransformContainer>>();
+        private static readonly ThreadSafeObjectPool<Stack<PushedState>> StateStackPool =
+            ThreadSafeObjectPool<Stack<PushedState>>.Default;
 
-        private Stack<PushedState> _states = StateStackPool.Count == 0 ? new Stack<PushedState>() : StateStackPool.Pop();
+        private static readonly ThreadSafeObjectPool<Stack<TransformContainer>> TransformStackPool =
+            ThreadSafeObjectPool<Stack<TransformContainer>>.Default;
 
-        private Stack<TransformContainer> _transformContainers = TransformStackPool.Count == 0
-            ? new Stack<TransformContainer>()
-            : TransformStackPool.Pop();
+        private Stack<PushedState> _states = StateStackPool.Get();
```

Comment:
So this PR also refactors the pooling for the `PushedState` stacks? Is this required by this feature or is it independent of the feature?